core/genesis.go

```diff
@@ -322,6 +316,19 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
 			statedb.SetState(addr, key, value)
 		}
 	}
+
+	// Configure any stateful precompiles that should activate in the genesis.
+	blockContext := types.NewBlockWithHeader(head)
+	err = ApplyPrecompileActivations(g.Config, nil, blockContext, statedb)
+	if err != nil {
+		panic(fmt.Sprintf("unable to configure precompiles in genesis block: %v", err))
+	}
+	// Configure any state upgrades that should activate in the genesis.
+	err = ApplyStateUpgrades(g.Config, nil, blockContext, statedb)
+	if err != nil {
+		panic(fmt.Sprintf("unable to configure state upgrades in genesis block: %v", err))
+	}
```

Comment:
What's the rationale for moving this here? Is this so that `ApplyStateUpgrades` occurs after setting the genesis allocation?

Reply:
Would it make sense to make it a requirement that state upgrades can only be enabled via `upgrade.json` and they cannot be used in the genesis since they should never be required there?

Reply:
We can do that, your 1st comment is correct. If we do that (disallow state upgrades in genesis), then we cannot combine the functions, which would you prefer?